packages/controller/lib/setup/setupInstall.js

```diff
@@ -640,266 +613,235 @@ function Install(options) {
 
         mime = mime || require('mime');
 
-        objects.getObject('system.adapter.' + adapter, (err, doc) => {
-            // Adapter is not installed - install it now
-            if (err || !doc || !doc.common.installedVersion) {
-                return that.installAdapter(adapter, () =>
-                    that.createInstance(adapter, options, callback));
-            }
+        let doc = await objects.getObjectAsync('system.adapter.' + adapter);
+        // Adapter is not installed - install it now
+        if (!doc || !doc.common.installedVersion) {
+            await this.installAdapterAsync(adapter);
+            doc = await objects.getObjectAsync('system.adapter.' + adapter);
+        }
 
-            // Check if some web pages should be uploaded
-            upload.uploadAdapter(adapter, true, false, () => {
-                upload.uploadAdapter(adapter, false, false, () => {
-                    objects.getObjectView('system', 'instance', {
-                        startkey: 'system.adapter.' + adapter + '.',
-                        endkey: 'system.adapter.' + adapter + '.\u9999'
-                    }, null, (err, res) => {
-                        objects.getObject('system.config', (err, systemConfig) => {
-                            const defaultLogLevel = systemConfig && systemConfig.common && systemConfig.common.defaultLogLevel;
-                            let a;
-                            if (err || !res) {
-                                console.error('host.' + hostname + ' error: view instanceStats ' + err);
-                                processExit(EXIT_CODES.CANNOT_READ_INSTANCES);
-                                return;
-                            }
+        // Check if some web pages should be uploaded
+        await upload.uploadAdapterAsync(adapter, true, false);
+        await upload.uploadAdapterAsync(adapter, false, false);
 
-                            // Count started instances
-                            if (doc.common.singleton && res.rows.length) {
-                                if (ignoreIfExists) {
-                                    callback && callback();
-                                    return;
-                                }
-                                console.error(`host.${hostname} error: this adapter does not allow multiple instances`);
-                                processExit(EXIT_CODES.NO_MULTIPLE_INSTANCES_ALLOWED);
-                                return;
-                            }
+        const res = await objects.getObjectViewAsync('system', 'instance', {
+            startkey: 'system.adapter.' + adapter + '.',
+            endkey: 'system.adapter.' + adapter + '.\u9999'
+        });
+        const systemConfig = await objects.getObjectAsync('system.config');
+        const defaultLogLevel = systemConfig && systemConfig.common && systemConfig.common.defaultLogLevel;
+        let a;
```

Comment:
Are not block scoped variabled better ? ;-) Just used in some for loops or?

Reply:
It is old code

Reply:
Changed